src/Messages/OWSMessageSender.m

```diff
@@ -700,36 +721,45 @@ - (void)groupSend:(NSArray<SignalRecipient *> *)recipients
-                if (!failedFuture.hasFailed) {
-                    // If at least one send succeeded, don't show message as failed.
-                    // Else user will tap-to-resend to all recipients, including those that already received the
-                    // message.
-                    return successHandler();
-                }
-            }
+        // failure from toc_thenAll yields an array of failed Futures, rather than the future's failure.
+        NSError *firstRetryableError = nil;
+        NSError *firstNonRetryableError = nil;
 
-            // At this point, all recipients must have failed.
-            // But we have all this verbose type checking because TOCFuture doesn't expose type information.
-            id lastError = errors.lastObject;
-            if ([lastError isKindOfClass:[TOCFuture class]]) {
-                TOCFuture *failedFuture = (TOCFuture *)lastError;
-                if (failedFuture.hasFailed) {
-                    id failureResult = failedFuture.forceGetFailure;
+        if ([failure isKindOfClass:[NSArray class]]) {
+            NSArray *groupSendFutures = (NSArray *)failure;
+            for (TOCFuture *groupSendFuture in groupSendFutures) {
+                if (groupSendFuture.hasFailed) {
+                    id failureResult = groupSendFuture.forceGetFailure;
                     if ([failureResult isKindOfClass:[NSError class]]) {
-                        // Generally we assume that failures are retryable
-                        return failureHandler((NSError *)failureResult, YES);
+                        NSError *error = failureResult;
+                        // Some errors should be ignored when sending messages
+                        // to groups.  See discussion on
+                        // NSError (OWSMessageSender) category.
+                        if ([error shouldBeIgnoredForGroups]) {
+                            continue;
+                        }
+
+                        if ([error isRetryable] && !firstRetryableError) {
+                            firstRetryableError = error;
+                        } else if (![error isRetryable] && !firstNonRetryableError) {
+                            firstNonRetryableError = error;
+                        }
                     }
                 }
             }
         }
 
-        DDLogWarn(@"%@ Unexpected generic failure: %@", self.tag, failure);
-        OWSAssert(NO);
-        return failureHandler(OWSErrorMakeFailedToSendOutgoingMessageError(), YES);
+        // If any of the group send errors are retryable, we want to retry.
+        // Therefore, prefer to propagate a retryable error.
+        if (firstRetryableError) {
+            return failureHandler(firstRetryableError);
+        } else if (firstNonRetryableError) {
+            return failureHandler(firstNonRetryableError);
+        } else {
+            // If we only received errors that we should ignore,
+            // consider this send a success.
```

Comment:
I don't think it's a big deal, but when you're in a group with *only* unregistered users, I expect we'll see the message successfully marked as "sent" even though we sent it to noone.

I don't think this is that big of a deal, since the sender will still see all the "unregistered user" warnings, they will likely not be that "surprised" that their message wasn't sent, but it's a little weird.

Would it be straight forward to do something like:
```
 if (message.sentRecipients.count == 0) {
    // do we need to mark it as non-retryable?
    NSError *error = OWSErrorMakeNoValidRecipientsError()
    failureHandler(errror)
}
```